internal/codeintel/bundles/persistence/cache/reader_cache.go

```diff
@@ -0,0 +1,149 @@
+	key    string
+	reader persistence.Reader
+	err    error
+	init   chan struct{}  // marks availability of reader field
+	once   sync.Once      // guards db.Close()
+	wg     sync.WaitGroup // concurrent use ref count
+}
+
+// New initializes a new cache with the given capacity and reader opener.
+func New(size int, opener ReaderOpener) *Cache {
+	return &Cache{
+		size:      size,
+		opener:    opener,
+		entries:   map[string]*list.Element{},
+		evictList: list.New(),
+	}
+}
+
+// WithReader calls the given function with a reader. If the reader has not yet initialized and does not
+// do so before context deadline, an error is returned. The reader initialization is not canceled due to
+// a context deadline here and will continue to run in the background until completion.
+func (c *Cache) WithReader(ctx context.Context, key string, fn CacheHandler) error {
+	entry := c.entry(key)
+	defer entry.wg.Done()
+
+	select {
+	case <-entry.init:
+		if entry.err != nil {
+			return entry.err
+		}
+		return fn(entry.reader)
+
+	case <-ctx.Done():
+		return ErrReaderInitializationDeadlineExceeded
+	}
+}
+
+// entry gets or creates a cache entry for the given key.
+func (c *Cache) entry(key string) *cacheEntry {
+	c.m.Lock()
+	defer c.m.Unlock()
+
+	if element, ok := c.entries[key]; ok {
+		entry := element.Value.(*cacheEntry)
+		entry.wg.Add(1)                  // Mark as in-use
+		c.evictList.MoveToFront(element) // Update recency data
+		return entry
+	}
+
+	entry := &cacheEntry{
+		key:  key,
+		init: make(chan struct{}),
+	}
+	entry.wg.Add(1)                         // Mark as in-use
+	element := c.evictList.PushFront(entry) // Update recency data
+	c.entries[key] = element
+
+	go func() {
+		// Mark as initialized
+		defer close(entry.init)
+
+		// Perform open procedure in a goroutine. We organize it this way so that
+		// if multiple differnet goroutines want to use the the same reader they do
+		// not all try to initialize it at once. In this situation, each goroutine
+		// will get a reference to the same entry and will block on the init channel
+		// being closed to signal that the handler cna be called.
+		entry.reader, entry.err = c.opener(key)
+		if entry.err != nil {
+			// Immediately remove from cache on error so that we do not permanently
+			// hold on to handles to missing bundles, as they may be uploaded in the
+			// future.
+			c.removeOnce(entry)
+		}
+	}()
+
+	// Delete n elements from the back of the list where n is how many elements over
+	// capacity we are.
+	n := len(c.entries) - c.size
+	for i, element := n, c.evictList.Back(); i > 0 && element != nil; i, element = i-1, element.Prev() {
+		c.removeOnce(element.Value.(*cacheEntry))
+	}
+
+	return entry
+}
+
+// removeOnce calls remove once in a goroutine.
+func (c *Cache) removeOnce(entry *cacheEntry) {
+	entry.once.Do(func() {
+		go func() { c.remove(entry) }()
+	})
+}
+
+// removeOnce blocks until the entry is initialized and exclusive access is available.
```

Comment:
```suggestion
// remove blocks until the entry is initialized and exclusive access is available.
```

Probable typo in the function doc.

Reply:
Oh I'm ALL ABOUT those.